go/vt/mysqlctl/backupengine.go

```diff
@@ -50,3 +56,84 @@ func GetBackupEngine() (BackupEngine, error) {
 	}
 	return be, nil
 }
+
+func findBackupToRestore(ctx context.Context, cnf *Mycnf, mysqld MysqlDaemon, logger logutil.Logger, dir string, bhs []backupstorage.BackupHandle, bm interface{}) (backupstorage.BackupHandle, error) {
+	var bh backupstorage.BackupHandle
+	var index int
+
+	for index = len(bhs) - 1; index >= 0; index-- {
+		bh = bhs[index]
+		rc, err := bh.ReadFile(ctx, backupManifest)
+		if err != nil {
+			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
+			continue
+		}
+
+		err = json.NewDecoder(rc).Decode(&bm)
+		rc.Close()
+		if err != nil {
+			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage (cannot JSON decode MANIFEST: %v)", bh.Name(), dir, err)
+			continue
+		}
+
+		logger.Infof("Restore: found backup %v %v to restore", bh.Directory(), bh.Name())
+		break
+	}
+	if index < 0 {
+		// There is at least one attempted backup, but none could be read.
+		// This implies there is data we ought to have, so it's not safe to start
+		// up empty.
+		return nil, errors.New("backup(s) found but none could be read, unsafe to start up empty, restart to retry restore")
+	}
+
+	// shutdown mysqld if it is running
+	// if we are retrying a failed restore, it is possible that mysqld failed to start
+	// so check for that
+	waitCtx, cancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+	defer cancel()
+	err := mysqld.Wait(waitCtx, cnf)
+	if err != nil || (waitCtx.Err() != nil && waitCtx.Err() == context.DeadlineExceeded) {
+		// do nothing
+	} else {
+		logger.Infof("Restore: shutdown mysqld")
+		// Starting from here we won't be able to recover if we get stopped by a cancelled
+		// context. Thus we use the background context to get through to the finish.
+		if err := mysqld.Shutdown(context.Background(), cnf, true); err != nil {
+			return nil, err
+		}
+	}
+	logger.Infof("Restore: deleting existing files")
+	if err := removeExistingFiles(cnf); err != nil {
+		return nil, err
+	}
+
+	logger.Infof("Restore: reinit config file")
+	if err := mysqld.ReinitConfig(context.Background(), cnf); err != nil {
+		return nil, err
+	}
+	return bh, nil
+}
+
+// open or create .restore file
+func openStateFile(cnf *Mycnf) (*os.File, error) {
```

Comment:
If we were to start writing to this file, then it wouldn't happen right away, it would be as and when we restore each file. But maybe that is unnecessary future-proofing. For the current implementation create / remove will work as well as open / close.

Reply:
Ah I see. It's just a sentinel file now, but in the future it might be a sort of checkpoint file? Maybe we should make this just a sentinel file, and if we need a checkpoint file later, it can be a separate file. That way the concerns are separated. What do you think?